src/lib.rs

```diff
@@ -0,0 +1,237 @@
+
+// mod mock;
+mod tests;
+
+/// Constants
+/////////////////////////////////////////////////////////////////
+
+const MAX_NUM_OF_PROPS_PER_CLASS: u16 = 30;
+
+/// Error messages for dispatchables
+const ERROR_NOT_IMPLEMENTED: &str = "Not implemented yet";
+
+const MAX_NAME_LENGTH: u16 = 100;
+const MAX_DESCRIPTION_LENGTH: u16 = 1000;
+
+pub type ClassId = u64;
+pub type EntityId = u64;
+
+/// Convenient composite time stamp 
+// #[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+// #[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
+// pub struct BlockAndTime<BlockNumber, Moment> {
+//     block: BlockNumber,
+//     time: Moment
+// }
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
+pub struct Class {
+    id: ClassId,
+
+    /// For Person, think "height", "weight", etc.
+    /// Unlikely to be more than roughly 20ish, often less.
+    properties: Vec<Property>,
+
+    /// All scehmas that are available for this class, think v0.0 Person, v.1.0 Person, etc.
+    schemas: Vec<ClassSchema>,
+
+    name: Vec<u8>,
+    description: Vec<u8>,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
+pub struct Entity {
+    id: EntityId,
+
+    /// The class id of this entity.
+    class_id: ClassId,
+
+    /// What schemas under which this entity of a class is available, think
+    /// v.2.0 Person schema for John , v3.0 Person schema for John
+    /// Unlikely to be more than roughly 20ish, assuming schemas for a given class eventually stableize, or that very old schema are eventually removed.
+    schemas: Vec<u16>, // indices of schema in corresponding class
+
+    /// Values for properties on class that are used by some schema used by this entity!
+    /// Length is no more than Class.properties.
+    values: Vec<(u16, PropertyValue)>, // Index is into properties vector of class. Fix anonymous type later.
+
+    name: Vec<u8>,
+    deleted: bool,
+}
+
+/// A schema defines what properties describe an entity
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
+pub struct ClassSchema {
+    version: u16,
+    properties: Vec<u8> // indices into properties vector for the corresponding class
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
+pub struct Property {
+    prop_type: PropertyType,
+    required: bool,
+    name: Vec<u8>,
+    description: Vec<u8>,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+pub enum PropertyType {
+    None,
+    Bool,
+    Uint16,
+    Uint32,
+    Uint64,
+    Int16,
+    Int32,
+    Int64,
+    Text(u32),
+    Internal(ClassId),
+    // External(ExternalProperty),
+}
+
+impl Default for PropertyType {
```

Comment:
Why is this needed?

Reply:
Why do we need a `Default` for `PropertyType`?
Because `PropertType` is an enum and if we want enum to be used in Substrate RPC that it should be serializable. This means that Substrate needs a default value for enums.